feat(zk): zksolc linking

Karrq opened the pull request:
# What :computer: 
* Add functions to use `zksolc --link`
* Process deploy-time linking (both `CREATE` and `CREATE2`)
* Update default zksolc to 1.5.9

# Why :hand:
* Allow tests & scripts to rely on contracts with libraries that have not been deployed yet

Karrq committed:
feat(zk): zksolc linking

Karrq committed:
fix(zk:libs): calculate addresses w/ proper nonce

chore: identify action to deploy libs in EraVM

Karrq committed:
Merge remote-tracking branch 'origin/main' into feat/zksolc-link

Karrq committed:
Merge remote-tracking branch 'origin/main' into feat/zksolc-link

Karrq committed:
fix: don't always assume DCC is present

Karrq committed:
feat(executor): `deploy_library` in strategy

Karrq committed:
fix(zk): create address computation

Karrq committed:
chore: cleanup unused imports

Karrq committed:
test(zk): deploy-time linking (script/test)

nbaztec commented on crates/zksync/compilers/src/link.rs:
```diff
@@ -0,0 +1,137 @@
+
+type LinkId = String;
+
+#[derive(Debug, Clone, Serialize, PartialEq, Eq, Hash)]
+#[serde(into = "String")]
+/// A library that zksolc will link against
```
```suggestion
/// A library that zksolc will link against
#[derive(Debug, Clone, Serialize, PartialEq, Eq, Hash)]
#[serde(into = "String")]
```

nbaztec commented on crates/zksync/compilers/src/link.rs:
```diff
@@ -0,0 +1,137 @@
+    }
+}
+
+#[derive(Debug, Clone, Serialize)]
+/// JSON Input for `zksolc link`
```
```suggestion
/// JSON Input for `zksolc link`
#[derive(Debug, Clone, Serialize)]
```

nbaztec commented on crates/zksync/compilers/src/link.rs:
```diff
@@ -0,0 +1,137 @@
+    }
+}
+
+#[derive(Debug, Clone, Deserialize)]
+/// JSON Output for `zksolc link`
+pub struct LinkJsonOutput {
```
```suggestion
/// JSON Output for `zksolc link`
#[derive(Debug, Clone, Deserialize)]
pub struct LinkJsonOutput {
```

nbaztec commented on crates/zksync/compilers/src/link.rs:
```diff
@@ -0,0 +1,137 @@
+    pub hash: String,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+/// Representation of a linked object given by zksolc
```
```suggestion
/// Representation of a linked object given by zksolc
#[derive(Debug, Clone, Deserialize)]
```

nbaztec commented on crates/zksync/compilers/src/link.rs:
```diff
@@ -0,0 +1,137 @@
+    pub libraries: HashSet<Library>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+/// Representation of a linked object given by zksolc
```
```suggestion
/// Representation of a linked object given by zksolc
#[derive(Debug, Clone, Deserialize)]
```

nbaztec commented on crates/zksync/core/src/lib.rs:
```diff
@@ -181,6 +181,20 @@ pub fn try_decode_create2(data: &[u8]) -> Result<(H256, H256, Vec<u8>)> {
 }
 
+/// Compute a CREATE address according to zksync
+pub fn compute_create_address(sender: Address, nonce: u64) -> Address {
```
https://github.com/matter-labs/zksync-era/blob/5a8818d93c66911dbe18ee81e404c28b4d933d1c/core/lib/multivm/src/versions/vm_m6/test_utils.rs#L161

Karrq committed:
chore: default zksolc to 1.5.8

Karrq commented on crates/evm/evm/src/executors/strategy.rs:
```diff
@@ -110,6 +127,13 @@ pub trait ExecutorStrategyExt {
+    ) -> Option<&'a mut DualCompiledContracts> {
+        None
+    }
+
```
Yes because it's not abstracted entirely into the strategy yet

elfedy commented on crates/config/src/zksync.rs:
```diff
@@ -204,7 +204,7 @@ pub fn config_create_project(
         zksolc
     } else if !config.offline {
-        let default_version = semver::Version::new(1, 5, 7);
+        let default_version = semver::Version::new(1, 5, 8);
```
It would be nice to have the version bump as a previous separate PR so that we have a clearer changelog for compiler versions. This feature is done thanks to that but it is not a requirement for the version bump per se right?

Karrq replied:
I will add that in the description, I had test failures because of 1.5.7

elfedy commented on crates/forge/src/multi_runner.rs:
```diff
@@ -508,11 +515,88 @@ impl MultiContractRunnerBuilder {
+                        .find(|(id, _)| id.source == needle.source && id.name == needle.name)
+                        .map(|(_, evm)| (needle, zk, evm))
+                })
+                .filter(|(_, zk, evm)| zk.bytecode.is_some() && evm.bytecode.is_some())
```
Shouldn't this be more of an assertion than a filter? Is it ok to keep on executing if we are missing some of the bytecodes?

elfedy replied:
Ok, I am just worried we would filter something we should be aware it's missing (for example what happened when using zksolc previous to `1.5.7` and trying to just deserialize the eravm field to get the bytecode) and then this fails later when trying to get the bytecode. If this cannot happen then we are good.

Karrq replied:
There are also some contracts which aren't meant to have bytecode like interfaces etc... Not sure the implications here tbh

elfedy commented on crates/forge/src/multi_runner.rs:
```diff
@@ -508,11 +515,88 @@ impl MultiContractRunnerBuilder {
+                        // FIXME: retrieve unlinked factory deps (1.5.9)
+                        zk_factory_deps: vec![zk_bytecode.to_vec()],
+                        evm_bytecode_hash: B256::from_slice(&keccak256(evm.as_ref())[..]),
+                        evm_deployed_bytecode: evm.to_vec(), // FIXME: is this ok? not really used
```
```suggestion
                        evm_deployed_bytecode: evm.to_vec(), // TODO(zk): is this ok? not really used
```

elfedy commented on crates/forge/src/multi_runner.rs:
```diff
@@ -508,11 +515,88 @@ impl MultiContractRunnerBuilder {
+
+        if let Some(dual_compiled_contracts) =
+            strategy.runner.zksync_get_mut_dual_compiled_contracts(strategy.context.as_mut())
+        {
```
Maybe this can be a `merge_linked_contracts` method on `DualCompiledContracts`

elfedy commented on crates/forge/tests/it/zk/linking.rs:
```diff
@@ -1,21 +1,47 @@
+    }
+);
+
+// FIXME(zk): add deploy-time linking to scripting
```
```suggestion
// TODO(zk): add deploy-time linking to scripting
```
What's the difference between FIXME and TODO? Is it worth it to add yet another distinctive word to the comments?

elfedy commented on crates/linking/src/lib.rs:
```diff
@@ -31,6 +42,18 @@ pub enum LinkerError {
     CyclicDependency,
 }
 
+/// Errors that can occur during linking.
```
Can we move this to a `zksync` module?

elfedy commented on crates/linking/src/lib.rs:
```diff
@@ -181,6 +204,60 @@ impl<'a> Linker<'a> {
+    /// When calling for `target` being an external library itself, you should check that `target`
+    /// does not appear in `libs_to_deploy` to avoid deploying it twice. It may happen in cases
+    /// when there is a dependency cycle including `target`.
+    pub fn zk_link_with_nonce_or_address(
```
Maybe we can define this as overrides as done in [config](https://github.com/matter-labs/foundry-zksync/blob/1b49d913be441cff8504432ff1c6da88e0f1ac28/crates/config/src/zksync.rs#L154). So we take `Linker` as the first arguments and define everything in a zksync module/file in order to keep this merge conflict free

Karrq replied:
moved to separate module & struct

elfedy commented on crates/linking/src/lib.rs:
```diff
@@ -265,13 +342,160 @@ impl<'a> Linker<'a> {
+
+        let zksolc = ZkSolcCompiler {
+            // NOTE(zk): zksolc --link --standard-json requires >1.5.8
+            // FIXME(zk): compiler from config
```
We should def use compiler from config and also error if we need linking and version < `.1.5.8` is used

Karrq replied:
We'd need to add the config to the `Linker` struct for that

Karrq replied:
I could make a `ZkLinker` and use it in the strategy anyways, that way we can do what we want, and pass in the config to the strategy's `link` to determine the compiler version

Karrq replied:
Done

elfedy commented on crates/strategy/zksync/src/executor.rs:
```diff
@@ -94,6 +121,76 @@ impl ExecutorStrategyRunner for ZksyncExecutorStrategyRunner {
+        // entirely in EraVM
+        let factory_deps = ctx.dual_compiled_contracts.fetch_all_factory_deps(dual_contract);
+
+        // persist existing paymaster data (needed?)
```
```suggestion
        // persist existing paymaster data (TODO: is this needed?)
```

elfedy commented on crates/zksync/compiler/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+//! # foundry-zksync
```
All definitions on this file were all moved to https://github.com/matter-labs/foundry-zksync/blob/1b49d913be441cff8504432ff1c6da88e0f1ac28/crates/config/src/zksync.rs#L154. Not sure how this file is still here, but maybe take the [current file](https://github.com/matter-labs/foundry-zksync/blob/main/crates/zksync/compilers/src/lib.rs) and add `pub mod link` there.

elfedy commented on crates/zksync/compilers/src/dual_compiled_contracts.rs:
```diff
@@ -133,10 +133,8 @@ impl DualCompiledContracts {
         let mut zksolc_all_bytecodes: HashMap<String, Vec<u8>> = Default::default();
         for (_, zk_artifact) in zk_output.artifacts() {
             if let (Some(hash), Some(bytecode)) = (&zk_artifact.hash, &zk_artifact.bytecode) {
-                // TODO: we can do this because no bytecode object could be unlinked
```
So compilation flow is now:
1) Build, get `DualCompiledContracts` but unlinked contracts won't be there.
2) Link using the output, add those to `DualCompliledContracts` then.
Is this correct?

Karrq replied:
Yep, linking without specifying any lib is not useful as the output remains the same

Karrq committed:
chore: lints

Karrq committed:
refactor: allow multiple lib deployments

Karrq committed:
refactor(link): move to executor strategy

fix(strategy): remove get_mut for `DualCompiledContracts`

Karrq committed:
fix: compilation

Karrq commented on crates/zksync/compilers/src/lib.rs:
```diff
@@ -7,6 +7,7 @@ pub mod artifacts;
 pub mod compilers;
 pub mod dual_compiled_contracts;
 pub mod libraries;
```
Not particularly, if not to help with backwards compatibility

elfedy replied:
We are removing `detect_missing_libraries` anyway, so imo we should just delete here

Karrq replied:
Well with my changes yeah, but should we be removing it?

elfedy replied:
Since the PR deletes the only place where this was being used (https://github.com/matter-labs/foundry-zksync/pull/800/files#diff-6bfe52bab6561edc6796c117419f60e77a296656bda5c199be3f3cf54354249eL411) then I think we should delete here. Either we keep both or delete both. I'm all in for deleting

Karrq replied:
Will do

Karrq committed:
feat(strategy:link): pass config

Karrq committed:
feat(zk:link): dedicated linker module

fix(zk:link): use version from config
refactor(zk:config): extract config -> zksolc compiler logic into
function
chore: lints

Karrq committed:
chore: more lints

chore: remove accidental file re-inclusion

Karrq committed:
feat(zk:link): version check

Karrq committed:
Merge remote-tracking branch 'origin/main' into feat/zksolc-link

Karrq committed:
chore: lints & fmt

Karrq committed:
chore: more formatting

Karrq committed:
fix(zk:link): retrieve factory dep hash

refactor(zk:link): avoid attempting to link fully-linked contracts

Karrq committed:
fix(zk:compilers): remove dead `libraries` module

Karrq committed:
feat(link:zk): create2 linking

feat(script:zk): link with zksolc

Karrq committed:
chore: formatting

Karrq committed:
feat(compiler:zk): `factory_dependencies_unlinked`

test(zk:link): use 1.5.9 for deploy-time linking
feat(zk:link): detect factory dependencies as link references for
library lookup
fix(zk:link): require 1.5.9 for deploy-time-linking

Karrq committed:
refactor: dual compiled contracts as map

Karrq committed:
feat(link:zk): recursive factory deps libs lookup

fix(link:zk): consistent artifact id prefixes
chore: formatting
refactor(link:zk): improve link targets filtering

Karrq committed:
fix(link:zk): invert bool

Karrq committed:
fix(artifacts:zk): `is_unlinked` logic

fix(link:zk): don't preemptively strip file prefixes
fix(link:zk): better factory dep detection
chore: formatting

Karrq committed:
feat(strategy): `deploy_library` CREATE2 mode

feat(strategy): return "broadcastable" tx
feat(link:zk): populate newly linked factory deps
feat(script): use `deploy_library`

Karrq committed:
fix(zk:transact): detect direct deployments

Karrq committed:
feat(script:zk): match CREATE/CREATE2 with EVM

Karrq committed:
fix(zk:libs): encode extra metadata

Karrq committed:
refactor(executors): `DeployLibResult`

fix(lib:zk): remove EVM lib deployments from broadcastable txs

Karrq committed:
fix(artifact:zk): avoid underflow

refactor(artifact:zk): `all_factory_deps` method
chore: formatting

Karrq committed:
fix(test:zk): proper stripping during link

feat(link:script): register target & deps

Jrigada commented on crates/linking/src/zksync.rs:
```diff
@@ -0,0 +1,468 @@
+            .into_iter()
+            // we strip these here because the file references are also relative
+            // and the linker wouldn't be able to properly detect matching factory deps
+            // (libraries are given separately and alredy stripped)
```
```suggestion
            // (libraries are given separately and already stripped)
```

Jrigada commented on crates/linking/src/zksync.rs:
```diff
@@ -0,0 +1,468 @@
+                        }
+
+                        targets.extend(ids); // queue factory deps for linking
+                        targets.push_back(id); // reque original target
```
```suggestion
                        targets.push_back(id); // requeue original target
```

Jrigada reviewed (commented):
Great job as always

Karrq commented on crates/evm/evm/src/executors/strategy.rs:
```diff
@@ -166,6 +241,127 @@ impl ExecutorStrategyRunner for EvmExecutorStrategyRunner {
+
+                // manually increase nonce when performing CALLs
+                executor
+                    .set_nonce(from, nonce + 1)
```
this was migrated from upstream's code, but the meaning is that EVM won't increase the nonce for just calls (only create or full tx, but here it's not a full tx), so foundry increments the nonce manually to account for that, since here it's a call to the deployer

Karrq commented on crates/evm/evm/src/executors/strategy.rs:
```diff
@@ -68,9 +109,32 @@ pub trait ExecutorStrategyRunner: Debug + Send + Sync + ExecutorStrategyExt {
         amount: U256,
     ) -> BackendResult<()>;
 
+    fn get_balance(&self, executor: &mut Executor, address: Address) -> BackendResult<U256>;
```
during zksync library deployment, I made our strategy to snapshot the nonce&balance before the lib deployment, to then restore them for EraVM, so the "co-deployment" happens with the same input parameters

Karrq commented on crates/zksync/core/src/lib.rs:
```diff
@@ -198,6 +199,38 @@ pub fn try_decode_create2(data: &[u8]) -> Result<(H256, H256, Vec<u8>)> {
+}
+
+/// Compute a CREATE2 address according to zksync
+pub fn compute_create2_address(
```
didn't find an equivalent for this unlinke create

Karrq commented on crates/zksync/core/src/vm/runner.rs:
```diff
@@ -51,7 +51,10 @@ where
     let nonce = ZKVMData::new(&mut ecx).get_tx_nonce(caller);
     let (transact_to, is_create) = match env.tx.transact_to {
-        TransactTo::Call(to) => (to.to_h160(), false),
+        TransactTo::Call(to) => {
```
had to change this one since transact may get deployments of the libs now, and we need to have the outcome parsed appropriately
technically speaking this is also helpful for broadcastable txs, as those could also contain deployments in this form

Karrq reviewed (commented):
Left comments to perhaps clarify some logic or choices

Karrq committed:
chore: fix spelling & docs

fix(test:zk): default to zksolc 1.5.9

Karrq committed:
fix(test:script): avoid expecting create2 output

fix(script): avoid unpacking create2 result
fix(script): avoid marking create2 result address as persistent

Karrq committed:
chore: clippy

Karrq committed:
chore: formatting

Karrq committed:
chore: codespell

Karrq committed:
fix(artifacts:zk): `is_unlinked` underflow

Karrq committed:
feat(compiler:zk): `objectFormat`

Karrq committed:
fix(script:link:zk): skip version check if no libs

Karrq committed:
fix(test:link:zk): avoid version check w/o libs

Karrq committed:
chore: clippy

Karrq committed:
fix(compiler:zk): optional object_format

Karrq committed:
fix(link:zk): ignore target version for lookup

Karrq committed:
Merge remote-tracking branch 'origin/main' into feat/zksolc-link

Karrq committed:
fix(link:zk): proper EVM deployed_bc/bc

Karrq committed:
chore: fmt

Karrq committed:
Merge remote-tracking branch 'origin/main' into feat/zksolc-link

nbaztec commented on crates/evm/evm/src/executors/strategy.rs:
```diff
@@ -48,6 +61,34 @@ pub struct ExecutorStrategy {
     pub context: Box<dyn ExecutorStrategyContext>,
 }
 
+pub struct LinkOutput {
```
Please do put this logic in its own module. Not convinces that it's currently required to be a separate trait - what are the aguments for it?

Karrq replied:
Just to make the distinction clearer and not "pollute" the strategy with linking-related items outside the linking stage

nbaztec commented on crates/forge/tests/it/test_helpers.rs:
```diff
@@ -612,10 +612,10 @@ pub async fn run_zk_script_test(
-        expected_broadcastable_txs
-    );
+    let txs = json["transactions"].as_array().expect("broadcastable txs");
+    tracing::debug!(?txs, "executed script");
```
Do we need the log in tests?

Karrq replied:
It's going to not be there most of the time, if not captured even by the test runner, but this way we can still have the information on what are the txs when stuff fails. I did add it for my own debugging initially but thought it could be useful overall

nbaztec commented on crates/linking/src/zksync.rs:
```diff
@@ -0,0 +1,466 @@
+
+        Ok(linked_artifacts.into_iter().collect())
+    }
+}
```
Wonder if we can add some unit tests to validate DFS and other heuristics in this module.

Karrq replied:
btw the DFS is from upstream, just slight edit to look thru factory deps too

Karrq replied:
Ugh, it's rather annoying to do unit test here because we depend on compilation output, and we don't have a way (AFAIK) to make something out of thin air there.
I'd need to refactor the algorithms to be independent of `self` or mock it out

nbaztec commented on crates/script/src/build.rs:
```diff
@@ -48,9 +47,198 @@ impl BuildData {
+            eyre::bail!("unable to link zk artifacts if no zk compilation output is provided");
+        };
+
+        let mut dual_compiled_contracts = self.dual_compiled_contracts.take().unwrap_or_default();
```
How do we have access to `dual_compiled_contracts` here, thought we moved them to strategy.

Karrq replied:
Script doesn't use strategies

Karrq committed:
chore: fmt

nbaztec commented on crates/strategy/zksync/src/executor/runner.rs:
```diff
@@ -73,6 +119,271 @@ impl ExecutorStrategyRunner for ZksyncExecutorStrategyRunner {
         Ok(())
     }
 
+    fn get_nonce(&self, executor: &mut Executor, address: Address) -> BackendResult<u64> {
```
As per https://github.com/matter-labs/foundry-zksync/pull/839 we should only use either `full_nonce`, `transaction_nonce`, or `deployment_nonce` within zksync context to disambiguate them.

Karrq replied:
This one is shared with EVM strategy, the zksync-specific one is aptly named

Karrq commented on crates/strategy/zksync/src/executor/runner.rs:
```diff
@@ -31,6 +53,23 @@ use crate::{
+        executor.backend.insert_account_storage(address, slot, new_full_nonce)?;
+
+        Ok(())
+    }
```
#839 takes care of after-execution state meanwhile this is meant for out-of-execution changes
(in this case to match the pre-deployment state with EVM, see https://github.com/matter-labs/foundry-zksync/pull/800#discussion_r1924030209)

Karrq replied:
(also it's not part of the trait interface because it's used internally only, hence the non-pub)

nbaztec commented on crates/zksync/compilers/src/dual_compiled_contracts.rs:
```diff
@@ -292,13 +297,56 @@ impl DualCompiledContracts {
+            name.map(|name| name == info.name.as_str()).unwrap_or(false)
+        });
+
+        full_matches.chain(path_matches).chain(name_matches).map(|(_, contract)| contract)
```
A unit test for this would be great

Karrq committed:
test(zk): `DualCompiledContracts::find` units

Karrq committed:
fix: clippy

Karrq committed:
refactor(strategy): dedicated linking module

Karrq committed:
fix: forgot to commit the new files

Karrq committed:
chore: fmt

Karrq committed:
Merge remote-tracking branch 'origin/main' into feat/zksolc-link

Karrq committed:
fix(zk:link): new nonce types

Karrq committed:
Merge remote-tracking branch 'origin/main' into feat/zksolc-link

elfedy commented on crates/forge/tests/it/zk/linking.rs:
```diff
@@ -44,8 +97,18 @@ forgetest_async!(
+    let zksolc = zksolc.unwrap_or_else(|| Version::new(1, 5, 9));
+    // force zksolc 1.5.9
+    let mut config = cmd.config();
+    if zksolc >= Version::new(1, 5, 9) {
```
we should not need to do this anymore because a) `1.5.10` has this as a warning instead. b) we won't be supporting `1.5.9`

Karrq replied:
yeah I can remove this and just go with 1.5.10

elfedy commented on crates/forge/tests/it/zk/linking.rs:
```diff
@@ -44,8 +97,18 @@ forgetest_async!(
     util::initialize(prj.root());
+
+    let zksolc = zksolc.unwrap_or_else(|| Version::new(1, 5, 9));
+    // force zksolc 1.5.9
```
where is the "forcing" happening?

Karrq replied:
If we don't specify a version we put _at least_ 1.5.9 for the test

elfedy commented on crates/script/src/build.rs:
```diff
@@ -48,9 +47,199 @@ impl BuildData {
         Linker::new(self.project_root.clone(), self.output.artifact_ids().collect())
     }
 
+    fn get_zk_linker(&self, script_config: &ScriptConfig) -> Result<ZkLinker<'_>> {
```
Maybe `build_data_get_linker` on a `zksync` module?

elfedy commented on crates/script/src/build.rs:
```diff
@@ -48,9 +47,199 @@ impl BuildData {
+    ///
+    /// If compilation for zksync is not enabled will return the
+    /// given EVM linked artifacts
+    async fn zk_link(
```
same

elfedy commented on crates/script/src/build.rs:
```diff
@@ -392,7 +577,7 @@ impl CompiledState {
         };
 
-        let linked_build_data = build_data.link_with_libraries(libraries)?;
+        let linked_build_data = build_data.link_with_libraries(&script_config, libraries).await?;
```
can we add a `NOTE(zk)` here if we are diverging from upstream?

elfedy commented on crates/script/src/runner.rs:
```diff
@@ -62,26 +64,28 @@ impl ScriptRunner {
         match libraries {
             ScriptPredeployLibraries::Default(libraries) => libraries.iter().for_each(|code| {
-                let result = self
+                let results = self
```
Is this diverging from upstream? can we add notes on how and why?

Karrq replied:
simply moved the upstream logic into `deploy_library` so we can override it with our strategy, but I can put a comment there yeah

Karrq committed:
Merge remote-tracking branch 'origin/main' into feat/zksolc-link

Karrq committed:
test(zk): use default zksolc version normally

Karrq committed:
docs: add notes on diverging sections

Karrq committed:
fix(zk): remove duplicate code from merge

Karrq committed:
refactor(script:zk): move linking to own module

Karrq committed:
chore: fmt

Karrq committed:
chore: clippy
